### PR TITLE
storage: adjust paddings for anaconda mode

### PR DESCRIPTION
**Pull request by KKoukiou**

### CODE REVIEW

**Comment by KKoukiou** [comment]
In anaconda mode storage details page should look like this:

**Comment by KKoukiou** [comment]
In anaconda mode storage details page should look like this:

breadcrumbs-above-details_1_

File: pkg/storaged/pages.jsx [hunk 1]
```diff
@@ -816,14 +816,14 @@
     return (
         <Page id="storage">
             { parent_crumbs.length > 0 &&
-            <PageBreadcrumb stickyOnBreakpoint={{ default: "top" }}>
+            <PageBreadcrumb {...(!client.in_anaconda_mode() ? { stickyOnBreakpoint: { default: "top" } } : {})}>
```
Comment (cockpituous) [inline pkg/storaged/pages.jsx:819]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240123-135148-fddbc75a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/pages.jsx [hunk 2]
```diff
@@ -816,14 +816,14 @@
...
                 <Breadcrumb>
                     { parent_crumbs }
                     <BreadcrumbItem isActive>{page_display_name(page)}</BreadcrumbItem>
                 </Breadcrumb>
             </PageBreadcrumb>
             }
-            <PageSection isFilled={false}>
+            <PageSection isFilled={false} padding={{ default: !client.in_anaconda_mode() > 0 ? "padding" : "noPadding" }}>
```
Comment (cockpituous) [inline pkg/storaged/pages.jsx:826]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240123-135148-fddbc75a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/storaged/pages.jsx:826]: That's right, that condition doesn't make sense. You probably want to just drop the `> 0`?

**Comment by jelly** [comment]
I find this a bit invasive if we even offer different CSS For anaconda. It would be great if this would include a pixel test for the anaconda case.

**Comment by KKoukiou** [comment]
> I find this a bit invasive if we even offer different CSS For anaconda. It would be great if this would include a pixel test for the anaconda case.

If you prefer this as a CSS overwrite instead I can also adjust. @jelly @mvollmer let me know your preference.

**Comment by mvollmer** [comment]
> > I find this a bit invasive if we even offer different CSS For anaconda. It would be great if this would include a pixel test for the anaconda case.
> 
> If you prefer this as a CSS overwrite instead I can also adjust. @jelly @mvollmer let me know your preference. Pixel tests are already there. See all failing tests in the storage scenario. :)

I am fine with this PR as it is, tweaking PF attributes with JavaScript.

File: pkg/storaged/pages.jsx (hunk 1 above)
Comment (mvollmer) [reply pkg/storaged/pages.jsx:819]: What does stickyOnBreakpoint actually do? It is not documented for PF 5.1.0, as far as I can see. Can we just remove it unconditionally? @garrett?

File: pkg/storaged/pages.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/storaged/pages.jsx:826]: Yes, and let's be positive: `client.in_anaconda_mode() ? "noPadding" : "padding"`

File: pkg/storaged/pages.jsx (hunk 1 above)
Comment (mvollmer) [reply pkg/storaged/pages.jsx:819]: Ah, sorry, this is `PageBreadcrumb` not `Breadcrumb`, which documents it. But I am still not sure what it actually does...

Comment (mvollmer) [reply pkg/storaged/pages.jsx:819]: (I was confused by what this does, but it seems to be necessary to control some padding below the breadcrumb.)

Can we write this as `stickyOnBreakpoint={ client.in_anaconda_mode() ? {} : {` default: false }}` ? That's less magic.

**Review by mvollmer — CHANGES_REQUESTED** [review]
See comments.

File: pkg/storaged/pages.jsx (hunk 1 above)
Comment (KKoukiou) [reply pkg/storaged/pages.jsx:819]: It makes that section sticky which has as a result visual effects like shadow bellow the bottom.

**Comment by mvollmer** [comment]
The pixel changes in mobile look unintended. Why did the dropdown button move?

File: pkg/storaged/pages.jsx [hunk 3]
```diff
@@ -816,14 +816,14 @@ export const StoragePage = ({ location, plot_state }) => {
     return (
         <Page id="storage">
             { parent_crumbs.length > 0 &&
-            <PageBreadcrumb stickyOnBreakpoint={{ default: "top" }}>
+            <PageBreadcrumb stickyOnBreakpoint={client.in_anaconda_mode() ? {} : { default: "top" }}>
```
Comment (cockpituous) [inline pkg/storaged/pages.jsx:819]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240124-094035-6c2858a4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/pages.jsx [hunk 4]
```diff
@@ -816,14 +816,14 @@ export const StoragePage = ({ location, plot_state }) => {
...
                 <Breadcrumb>
                     { parent_crumbs }
                     <BreadcrumbItem isActive>{page_display_name(page)}</BreadcrumbItem>
                 </Breadcrumb>
             </PageBreadcrumb>
             }
-            <PageSection isFilled={false}>
+            <PageSection isFilled={false} padding={{ default: client.in_anaconda_mode() ? "noPadding" : "padding" }}>
```
Comment (cockpituous) [inline pkg/storaged/pages.jsx:826]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240124-094035-6c2858a4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by mvollmer — DISMISSED** [review]
Thanks!

File: pkg/storaged/pages.jsx [hunk 5]
```diff
@@ -814,16 +814,16 @@ export const StoragePage = ({ location, plot_state }) => {
     }
 
     return (
-        <Page id="storage">
+        <Page id="storage" className={client.in_anaconda_mode() ? "anaconda-mode" : ""}>
```
Comment (cockpituous) [inline pkg/storaged/pages.jsx:817]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240124-111158-08af2ed3-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/pages.jsx [hunk 6]
```diff
@@ -814,16 +814,16 @@ export const StoragePage = ({ location, plot_state }) => {
...
             { parent_crumbs.length > 0 &&
-            <PageBreadcrumb stickyOnBreakpoint={{ default: "top" }}>
+            <PageBreadcrumb stickyOnBreakpoint={client.in_anaconda_mode() ? {} : { default: "top" }}>
```
Comment (cockpituous) [inline pkg/storaged/pages.jsx:819]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240124-111158-08af2ed3-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/pages.jsx [hunk 7]
```diff
@@ -814,16 +814,16 @@ export const StoragePage = ({ location, plot_state }) => {
...
                 <Breadcrumb>
                     { parent_crumbs }
                     <BreadcrumbItem isActive>{page_display_name(page)}</BreadcrumbItem>
                 </Breadcrumb>
             </PageBreadcrumb>
             }
-            <PageSection isFilled={false}>
+            <PageSection isFilled={false} padding={client.in_anaconda_mode() ? { default: "noPadding" } : {}}>
```
Comment (cockpituous) [inline pkg/storaged/pages.jsx:826]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240124-111158-08af2ed3-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/storage.scss [hunk 8]
```diff
@@ -373,3 +373,13 @@ a.disabled {
     // 1em: margin between text and progress bar
     padding-inline-end: calc(var(--pf-v5-global--spacer--sm) + 3.5em + 1em) !important;
 }
+
+// FIXME: this makes the first card in anaconda mode inconsistent with other cards which have shadow
```
Comment (mvollmer) [inline pkg/storaged/storage.scss:377]: Could you put a class on just the breadcrumb to tweak its shadow? Would that be more proper? I am afraid I don't understand what is improper about your solution here.

Comment (mvollmer) [reply pkg/storaged/storage.scss:377]: Also, what about changing the breadcrumb visuals also in non-Anaconda mode to match what we are trying to achieve here?

Comment (KKoukiou) [reply pkg/storaged/storage.scss:377]: I found a way to control box shadow in all directions. I think I can fix it properly.

Comment (KKoukiou) [reply pkg/storaged/storage.scss:377]: I thought that box-shadow it not possible to control per-direction - but apparently it is. I will utilize this and we should be good.

Comment (KKoukiou) [reply pkg/storaged/storage.scss:377]: Ok it's hard. I suggest we get this as is in, unless @mvollmer has a better idea, and we can consider how we can improve it in the future?

File: pkg/storaged/pages.jsx (hunk 5 above)
Comment (cockpituous) [inline pkg/storaged/pages.jsx:817]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240125-115723-d7fb71dd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/pages.jsx [hunk 9]
```diff
@@ -814,16 +814,16 @@
     }
 
     return (
-        <Page id="storage">
+        <Page id="storage" className={client.in_anaconda_mode() ? "anaconda-mode" : ""}>
             { parent_crumbs.length > 0 &&
-            <PageBreadcrumb stickyOnBreakpoint={{ default: "top" }}>
+            <PageBreadcrumb stickyOnBreakpoint={client.in_anaconda_mode() ? {} : { default: "top" }}>
```
Comment (cockpituous) [inline pkg/storaged/pages.jsx:819]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240125-115723-d7fb71dd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/pages.jsx [hunk 10]
```diff
@@ -814,16 +814,16 @@
...
                 <Breadcrumb>
                     { parent_crumbs }
                     <BreadcrumbItem isActive>{page_display_name(page)}</BreadcrumbItem>
                 </Breadcrumb>
             </PageBreadcrumb>
             }
-            <PageSection isFilled={false}>
+            <PageSection isFilled={false} padding={client.in_anaconda_mode() ? { default: "noPadding" } : {}}>
```
Comment (cockpituous) [inline pkg/storaged/pages.jsx:826]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240125-115723-d7fb71dd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/storage.scss [hunk 11]
```diff
@@ -373,3 +373,13 @@ a.disabled {
...
+// For fixing this propertly we would need to refactor the whole page rendering for anaconda mode
+.anaconda-mode {
+    .pf-v5-c-page__main-breadcrumb + .pf-v5-c-page__main-section {
+        .pf-v5-c-card {
+            box-shadow: none;
+        }
+    }
```
Comment (garrett) [inline pkg/storaged/storage.scss:384]: Would something like this work better? It visually merges the card using the PF margins and should handle mobile sizes too. Although we probably want to find out where that 576px magic number is coming from and use the SASS variable instead. (Sadly, CSS variables are not supported in media breakpoints yet.)

```suggestion
    .pf-v5-c-page__main-breadcrumb {
        margin: var(--pf-v5-c-page__main-section--PaddingTop) var(--pf-v5-c-page__main-section--PaddingRight) calc(-2 * var(--pf-v5-c-page__main-section--PaddingTop)) var(--pf-v5-c-page__main-section--PaddingLeft);
        // Remove the bottom of PF's box-shadow, but leave the top and sides intact.
        clip-path: rect(-10px -10px 100% calc(100% + 10px));
  
        @media screen and (max-width: 576px) {
              margin-inline: 0;
        }
    }
```

File: pkg/storaged/storage.scss (hunk 8 above)
Comment (mvollmer) [reply pkg/storaged/storage.scss:377]: There is `hasShadowBottom`, could we set that to false?

**Comment by mvollmer** [comment]
Alright, I have added a breadcrumb to the first card of a page (in Anaconda mode):

image

I can't see any artifacts.

**Comment by mvollmer** [comment]
@KKoukiou, let's do some mutal review here, and when you are happy with my changes, I'll approve formally.

File: pkg/storaged/pages.jsx [hunk 12]
```diff
@@ -802,28 +826,14 @@ export const StorageDescription = ({ title, value, action, children }) => {
...
             <PageBreadcrumb stickyOnBreakpoint={{ default: "top" }}>
-                <Breadcrumb>
-                    { parent_crumbs }
-                    <BreadcrumbItem isActive>{page_display_name(page)}</BreadcrumbItem>
-                </Breadcrumb>
+                <StorageBreadcrumb page={page} />
             </PageBreadcrumb>
             }
-            <PageSection isFilled={false}>
+            <PageSection isFilled={false} padding={client.in_anaconda_mode() ? { default: "noPadding" } : {}}>
```
Comment (cockpituous) [inline pkg/storaged/pages.jsx:836]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240131-150341-67c3f528-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/storage.scss (hunk 8 above)
Comment (mvollmer) [reply pkg/storaged/storage.scss]: > There is `hasShadowBottom`, could we set that to false?

No, it is already false by default.

File: pkg/storaged/pages.jsx (hunk 12 above)
Comment (cockpituous) [inline pkg/storaged/pages.jsx:836]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19876-20240201-080923-67c3f528-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by KKoukiou — COMMENTED** [review]
Looks great, thanks for fixing that without complicated CSS :)

**Review by mvollmer — APPROVED** [review]

**Pull request merged**